v2 default baselines for projects/accounts STO-7045 (#5163)
harness/developer-hub: docs/security-testing-orchestration/use-sto/set-up-sto-pipelines/set-up-baselines.md

---
title: Target baselines in STO
description: How to set up baselines. Every scanned target needs one.  
sidebar_label: Target baselines
sidebar_position: 50
---

Every scanned target should have a _baseline_, which represents the root variant such as the `main` branch or the `latest` tag.  Baselines make it easy to identify issues in the baseline vs. issues in a downstream variant derived from that baseline. 

:::info key concepts: target baselines
- [Every scanned target needs a baseline](#every_target_needs_a_baseline) to enable the full suite of STO features.
- You can specify a baseline using a fixed value or a regular expression (RegEx). 

> - When a target with a RegEx baseline gets scanned, and the variant name matches the RegEx, the target baseline updates automatically. [RegEx baselines are recommended](#advantages_of_regex_baselines) unless you know that the [target baseline is fixed](#specify-a-fixed-baseline-for-an-individual-target) and will not change. 

Reviewer: It feels a bit off to say that RegEx baselines "are recommended, except". Rather, I'd call out the reason to use each of the two options with an example. Such as:
- If you always want to compare against a specific variant--such as the `main` branch of your git repository--use a fixed value...
- If the variant you want to use for comparison will change--for example, with each release a new `release-X.Y.Z` branch is created, and you want to compare to that--use a RegEx baseline...

> - You can specify [default RegEx baselines](#default-regular-expressions-for-target-baselines) for each target type.

Reviewer: ```suggestion
- You can specify a [default RegEx baseline](#default-regular-expressions-for-target-baselines) for each target type. This RegEx will automatically be used as the baseline for any new targets.
```

- Repositories and images have predefined defaults:
  - For repositories, the default is the `main` or `master` branch.

>   - For images, the default is `latest` or the most recently scanned two-dot release. 

Reviewer: Combine with the above bullet, and use "Semantic Versioned release" instead of "two-dot release"

-->

To view and edit the current baseline for a target, go to **Test Targets**:

![](../../get-started/static/targets-and-baselines.png)

<a name="every_target_needs_a_baseline"></a>

<details>

<summary>Every target needs a baseline </summary>

Every target needs a baseline to enable the full suite of STO features. Here's why:

  - For developers, it’s critical to distinguish between security issues in the baseline vs. issues in the variant you’re working on. You want to detect and resolve issues in your downstream branch (the variant) before merging, so you don’t introduce them into the main branch (the baseline). 

  - When your target has a baseline defined, the [Security Tests tab](//docs/security-testing-orchestration/use-sto/view-and-troubleshoot-vulnerabilities/view-scan-results) divides your issues into two lists:

    - **Only in \<_target_>:\<_variant_>** Issues detected in the scanned variant only.

    - **Common to \<_target_>:\<_baseline_>** Issues also detected in the target baseline.

>   - While the **Security Tests** tab highlights downstream issues, the [STO Overview](/docs/security-testing-orchestration/use-sto/view-and-troubleshoot-vulnerabilities/sto-overview) and [Security Testing Dashboard](/docs/security-testing-orchestration/use-sto/view-and-troubleshoot-vulnerabilities/security-testing-dashboard) focus on baseline issues. These views enable security personnel and other non-developers to monitor, investigate, and address issues in production-ready targets and view vulnerability trends over time.  

Reviewer: ```suggestion
  - While the **Security Tests** tab highlights downstream issues, the [STO Overview](/docs/security-testing-orchestration/use-sto/view-and-troubleshoot-vulnerabilities/sto-overview) and [Security Testing Dashboard](/docs/security-testing-orchestration/use-sto/view-and-troubleshoot-vulnerabilities/security-testing-dashboard) only show baseline issues. These views enable security personnel and other non-developers to monitor, investigate, and address issues in production-ready targets and view vulnerability trends over time.  
```

  - In short, baselines make it easy for developers to drill down into “shift-left” issues in downstream variants and security personnel to drill down into “shift-right” issues in  production targets.
  
</details>

<a name="advantages_of_regex_baselines"></a>
<details>
<summary>Advantages of using regular expressions to define baselines</summary>

> Defining your baselines using regular expressions provides significant benefits over using fixed values such as `main` or `latest`. 

Reviewer: Again, there isn't really a benefit that comes with RegEx baselines. It's a matter of if the customer's baseline needs to be changed automatically over time, or not.

   - A new release branch `1.2.3` is created. You scan this branch, which matches your RegEx, and it becomes the current baseline. 
   - The scan detects a new vulnerability, **vXYZ**, with a severity of MEDIUM.
   - A few sprints later, the vulnerability is remediated in a branch that gets merged into `1.2.7`.
   - `1.2.7` is scanned before getting merged into `main`, and **vXYZ** is no longer in the scan results. 

   ![](../static/dynamic-baselines-example.png)

</details>
 
### Specify fixed baselines in STO

You can specify a fixed value if you know that the baseline for a target will not change. 

1. Scan the target with the baseline variant if you haven't yet done so.

2. Go to **Test Targets**.

3. In **Baseline for Comparison**, set the value type to **Fixed** and select the baseline from the menu of scanned variants. 

   <DocImage path={require('./static/baseline-fixed-individual-target.png')} width="100%" height="100%" title="Add shared path for scan results" /> 

### Specify RegEx baselines in STO

Suppose your organization publishes container images with a `latest` tag as well as two-dot versions such as:  `3.0.0` , `3.17`, `3.17.3`, `3.18.0`, `3.18.2`, ...

> With this cadence, you can create a RegEx baseline that updates whenever you create a new image with a matching tag. If you scan a new image target, `jdoe/myimage`, with the following tags, the baseline gets updated as follows:

Reviewer: ```suggestion
To support proper functioning of STO with this workflow, you can set a RegEx baseline that updates whenever you scan a new image with a matching tag. If you scan a new image target, `jdoe/myimage`, with the following tags, the baseline gets updated as follows:
```

Reviewer: `jdoe/myimage` doesn't match the below example, but I'm also not clear on what the below example is trying to show. Perhaps using `3.2.0` and `3.3.0` as the example image tags would be clearer?

Author: Resolved in final commit
